csrc/cpu/cpu_types_arm.hpp

```diff
@@ -49,432 +53,344 @@ constexpr void unroll_loop(F&& f) {
+    reg.val[0] = Vectorized<c10::BFloat16>::loadu(base + 0);
+    reg.val[1] = Vectorized<c10::BFloat16>::loadu(base + 8);
+  };
 
-  explicit BF16Vec16(bfloat16x8x2_t data) : reg(data) {};
+#ifdef ARM_BF16_SUPPORT
+  explicit BF16Vec16(bfloat16x8x2_t data) : reg{data.val[0], data.val[1]} {};
+#endif
 
   explicit BF16Vec16(const FP32Vec16&);
 
   explicit BF16Vec16(float32x4x4_t v)
-      : reg({vcvtq_high_bf16_f32(vcvtq_low_bf16_f32(v.val[0]), v.val[1]),
-             vcvtq_high_bf16_f32(vcvtq_low_bf16_f32(v.val[2]), v.val[3])}) {};
+      : reg({convert_float_bfloat16(v.val[0], v.val[1]),
+             convert_float_bfloat16(v.val[2], v.val[3])}) {};
+
+  void save(void* ptr) const {
+    c10::BFloat16* base = reinterpret_cast<c10::BFloat16*>(ptr);
+    reg.val[0].store(base + 0);
+    reg.val[1].store(base + 8);
+  };
 
-  void save(void* ptr) const { *reinterpret_cast<bfloat16x8x2_t*>(ptr) = reg; };
   void save(void* ptr, const int elem_num) const {
-    int full_blocks = elem_num / NUM_ELEMENTS_REG(reg.val[0]);
-    int remainder = elem_num % NUM_ELEMENTS_REG(reg.val[0]);
+    c10::BFloat16* base = reinterpret_cast<c10::BFloat16*>(ptr);
+    size_t num_elements = reg.val[0].size();
+
+    int full_blocks = elem_num / num_elements;
+    int remainder = elem_num % num_elements;
     for (int i = 0; i < full_blocks; i++)
-      vst1q_bf16(
-          reinterpret_cast<__bf16*>(ptr) + NUM_ELEMENTS_REG(reg.val[0]) * i,
-          reg.val[i]);
+      reg.val[i].store(base + num_elements * i);
+
     if (remainder > 0) {
-      bfloat16x8_t temp = reg.val[full_blocks];
-      bfloat16_t* base = reinterpret_cast<bfloat16_t*>(ptr) + full_blocks * 8;
-      if (remainder > 0) base[0] = vgetq_lane_bf16(temp, 0);
-      if (remainder > 1) base[1] = vgetq_lane_bf16(temp, 1);
-      if (remainder > 2) base[2] = vgetq_lane_bf16(temp, 2);
-      if (remainder > 3) base[3] = vgetq_lane_bf16(temp, 3);
-      if (remainder > 4) base[4] = vgetq_lane_bf16(temp, 4);
-      if (remainder > 5) base[5] = vgetq_lane_bf16(temp, 5);
-      if (remainder > 6) base[6] = vgetq_lane_bf16(temp, 6);
+      reg.val[full_blocks].store(base + num_elements * full_blocks, remainder);
     }
   };
 };
 
 struct BF16Vec32 : public Vec<BF16Vec32> {
   constexpr static int VEC_ELEM_NUM = 32;
 
-  bfloat16x8x4_t reg;
-
-  explicit BF16Vec32(const void* ptr)
-      : reg(*reinterpret_cast<const bfloat16x8x4_t*>(ptr)) {};
-
-  explicit BF16Vec32(bfloat16x8x4_t data) : reg(data) {};
+  Vec4x8bf reg;
 
+  explicit BF16Vec32(const void* ptr) {
+    const c10::BFloat16* base = reinterpret_cast<const c10::BFloat16*>(ptr);
+    reg.val[0] = Vectorized<c10::BFloat16>::loadu(base + 0);
+    reg.val[1] = Vectorized<c10::BFloat16>::loadu(base + 8);
+    reg.val[2] = Vectorized<c10::BFloat16>::loadu(base + 16);
+    reg.val[3] = Vectorized<c10::BFloat16>::loadu(base + 24);
+  }
+#ifdef ARM_BF16_SUPPORT
+  explicit BF16Vec32(bfloat16x8x4_t data)
```

Comment:
can we ditch this constructor taking `bfloat16x8x4_t` as input and remove the `#ifdef ARM_BF16_SUPPORT`